tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -0,0 +1,235 @@
+/*
+ * Parts derived from tests/kernel/fatal/src/main.c, which has the
+ * following copyright and license:
+ *
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <ztest.h>
+#include <kernel_structs.h>
+#include <string.h>
+#include <stdlib.h>
+
+#define INFO(fmt, ...) printk(fmt, ##__VA_ARGS__)
+
+/* ARM is a special case, in that k_thread_abort() does indeed return
+ * instead of calling _Swap() directly. The PendSV exception is queued
+ * and immediately fires upon completing the exception path; the faulting
+ * thread is never run again.
+ */
+#ifndef CONFIG_ARM
+FUNC_NORETURN
+#endif
+void _SysFatalErrorHandler(unsigned int reason, const NANO_ESF *pEsf)
+{
+	INFO("Caught system error -- reason %d\n", reason);
+	ztest_test_pass();
```

Comment:
consider checking that the induced error is the expected error state, and not some unrelated fatal error

Reply:
> consider checking that the induced error is the expected error state, and not some unrelated fatal error

On x86, the reason will differ for different tests (6 for most of them, 7 for the ones that trigger an oops), so I'd have to set the expected reason value in a global in each test function prior to triggering the fault/oops and then compare it here before calling ztest_test_pass(). On arm, it appears that reason is not set (just 0), at least based on what I see with the protection test suite, so it would be arch-specific.  Is it worth adding the checking per-arch and per-test?

Reply:
Sorry, I see that 0 is _NANO_ERR_HW_EXCEPTION for arm, used for mpu/bus/usage faults.  So we could still distinguish the oops cases from the fault cases, I suppose.  But we'd essentially need to define two expected values, one for fault and one for oops, on a per-arch basis, and then set it in each test function.

Reply:
Let's punt on this for now, the NanoFatalErrorHandler stuff is due for a major overhaul and we can re-visit once that lands.